presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/operator/aggregation/GeometryStateFactory.java

```diff
@@ -0,0 +1,148 @@
+
+import com.esri.core.geometry.ogc.OGCGeometry;
+import com.facebook.presto.array.ObjectBigArray;
+import com.facebook.presto.geospatial.GeometryType;
+import com.facebook.presto.operator.aggregation.state.AbstractGroupedAccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorStateFactory;
+
+/**
+ * A factory for creating single and grouped forms of the {@link GeometryState}
+ */
+public class GeometryStateFactory
+        implements AccumulatorStateFactory<GeometryState>
+{
+    @Override
+    public GeometryState createSingleState()
+    {
+        return new SingleGeometryState();
+    }
+
+    @Override
+    public Class<? extends GeometryState> getSingleStateClass()
+    {
+        return SingleGeometryState.class;
+    }
+
+    @Override
+    public GeometryState createGroupedState()
+    {
+        return new GroupedGeometryState();
+    }
+
+    @Override
+    public Class<? extends GeometryState> getGroupedStateClass()
+    {
+        return GroupedGeometryState.class;
+    }
+
+    public static class GroupedGeometryState
+            extends AbstractGroupedAccumulatorState
+            implements GeometryState
+    {
+        private final ObjectBigArray<OGCGeometry> array = new ObjectBigArray<>();
+        private long size;
+
+        @Override
+        public OGCGeometry getGeometry()
+        {
+            return array.get(getGroupId());
+        }
+
+        @Override
+        public void setGeometry(OGCGeometry geometry)
+        {
+            final OGCGeometry previousValue = this.array.get(getGroupId());
+            size -= getGeometryMemorySize(previousValue);
+            size += getGeometryMemorySize(geometry);
+            this.array.set(getGroupId(), geometry);
+        }
+
+        @Override
+        public void union(OGCGeometry geometry)
+        {
+            final OGCGeometry previousValue = this.array.get(getGroupId());
+            size -= getGeometryMemorySize(previousValue);
+            final OGCGeometry newValue = previousValue == null ? geometry : previousValue.union(geometry);
+            this.array.set(getGroupId(), newValue);
+            size += getGeometryMemorySize(newValue);
+        }
+
+        @Override
+        public void ensureCapacity(long size)
+        {
+            array.ensureCapacity(size);
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return size + array.sizeOf();
+        }
+    }
+
+    // Do a best-effort attempt to estimate the memory size
+    private static long getGeometryMemorySize(OGCGeometry geometry)
+    {
+        if (geometry == null) {
+            return 0;
+        }
+        // TODO: The following geometries will throw an NPE on the current version of Esri geometry.
+        // Once a release is cut that contains this PR we can remove this check:
+        // https://github.com/Esri/geometry-api-java/pull/193
+        switch (GeometryType.getForEsriGeometryType(geometry.geometryType())) {
```

Comment:
perhaps, skip this check for non-empty geometries

```
if (geometry.isEmpty()) {
    return 0;
}

return geometry.estimateMemorySize();
```

Reply:
I'll use `24`, as I'd worry about a scenario where one has a large grouped aggregation that consists of a non-trivial number of empty geometries.